src/app/backend/auth/kubeconfig.go

```diff
@@ -14,4 +14,105 @@
 
 package auth
 
-// TODO(floreks) implement kubeconfig based authenticator
+import (
+	"errors"
+
+	authApi "github.com/kubernetes/dashboard/src/app/backend/auth/api"
+	yaml "gopkg.in/yaml.v2"
+	"k8s.io/client-go/tools/clientcmd/api"
+)
+
+type contextInfo struct {
+	User string `yaml:"user"`
+}
+
+type contextEntry struct {
+	Name    string      `yaml:"name"`
+	Context contextInfo `yaml:"context"`
+}
+
+type userEntry struct {
+	Name string   `yaml:"name"`
+	User userInfo `yaml:"user"`
+}
+
+type userInfo struct {
+	Token    string `yaml:"token"`
+	Username string `yaml:"username"`
+	Password string `yaml:"password"`
+}
+
+type kubeConfig struct {
+	Contexts       []contextEntry `yaml:"contexts"`
+	CurrentContext string         `yaml:"current-context"`
+	Users          []userEntry    `yaml:"users"`
+}
+
+// Implements Authenticator interface
+type kubeConfigAuthenticator struct {
+	fileContent []byte
+}
+
+// GetAuthInfo implements Authenticator interface. See Authenticator for more information.
+func (self *kubeConfigAuthenticator) GetAuthInfo() (api.AuthInfo, error) {
+	kubeConfig, err := self.parseKubeConfig(self.fileContent)
+	if err != nil {
+		return api.AuthInfo{}, err
+	}
+
+	info, err := self.getCurrentUserInfo(*kubeConfig)
+	if err != nil {
+		return api.AuthInfo{}, err
+	}
+
+	return self.getAuthInfo(info)
+}
+
+func (self *kubeConfigAuthenticator) parseKubeConfig(bytes []byte) (*kubeConfig, error) {
+	kubeConfig := new(kubeConfig)
+	if err := yaml.Unmarshal(bytes, kubeConfig); err != nil {
+		return nil, err
+	}
+
+	return kubeConfig, nil
+}
+
+func (self *kubeConfigAuthenticator) getCurrentUserInfo(config kubeConfig) (userInfo, error) {
+	userName := ""
+	for _, context := range config.Contexts {
+		if context.Name == config.CurrentContext {
+			userName = context.Context.User
+		}
+	}
+
+	if len(userName) == 0 {
+		return userInfo{}, errors.New("Context matching current context not found. Check if your config file is valid.")
+	}
+
+	for _, user := range config.Users {
+		if user.Name == userName {
+			return user.User, nil
+		}
+	}
+
+	return userInfo{}, errors.New("User matching current context user not found. Check if your config file is valid.")
+}
+
+func (self *kubeConfigAuthenticator) getAuthInfo(info userInfo) (api.AuthInfo, error) {
+	if len(info.Token) == 0 && (len(info.Password) == 0 || len(info.Username) == 0) {
+		return api.AuthInfo{}, errors.New("Not enough data to create auth info structure.")
+	}
+
+	return api.AuthInfo{
+		Username: info.Username,
+		Password: info.Password,
+		Token:    info.Token,
+	}, nil
+}
+
+// NewBasicAuthenticator returns Authenticator based on LoginSpec.
+func NewKubeConfigAuthenticator(spec *authApi.LoginSpec) authApi.Authenticator {
+	return &kubeConfigAuthenticator{
+		fileContent: []byte(spec.KubeConfig),
+	}
+}
```

Comment:
I like the look of this file, 2 questions, 
1) might be mostly me being not up to speed. but does this 'authentication method' chose which to use based on the 'enabledness' of Token and Basic authentication? and if so, if both are enabled, does it have a preference?

2) do we want to support some auth-provider type kubeconfig files? e.g. GKE writes something like:
```
- name: AzureDiamond
  user:
    auth-provider:
      config:
        access-token: hunter2
        cmd-args: config config-helper --format=json
        cmd-path: /google/google-cloud-sdk/bin/gcloud
        expiry: 2017-08-31 17:19:02
        expiry-key: '{.credential.token_expiry}'
        token-key: '{.credential.access_token}'
      name: gcp
```
and while we ofcourse can't do the token refresh by running gcloud, the access-token in the config is a valid bearer token. 

Both are mostly informational questions, I'm not saying we should do fancy stuff for auth-providers, just wondering if it might be worth the effort

Reply:
ow, I should read your description for question 1. still wondering about second part of that question though

Reply:
1. I'll add check which auth modes are enabled and allow only user fields that match them. In example if only token mode will be enabled then kubeconfig authenticator won't allow to log in with username/password.

2. We are planning to add support for external identity providers in the future. Right now I am not sure if this will be done through kubeconfig or similar to popular `login with facebook/google` options. For now we'll stick to basic options. Best option for us would be some API exposed by apiserver that would handle most of this for us.